### ISSUE TITLE
JOIN ON t1.k IS NOT DISTINCT FROM t2.k does not translate to an equi-join

### ISSUE THREAD

**Issue by yonillasky** [prestodb/presto#10900]
The following query:
```
select * 
from (
    SELECT * FROM (VALUES ('a', 1), ('b', 2), (NULL, 3)) AS t(k, v)
) t1
join (
    SELECT * FROM (VALUES ('a', 4), ('b', 5), (NULL, 6)) AS t(k, v)
) t2 
ON t1.k IS NOT DISTINCT FROM t2.k
```
generates a plan that has a CrossJoin operator in it.

Whereas, the query:
```
select * 
from (
    SELECT * FROM (VALUES ('a', 1), ('b', 2), (NULL, 3)) AS t(k, v)
) t1
join (
    SELECT * FROM (VALUES ('a', 4), ('b', 5), (NULL, 6)) AS t(k, v)
) t2 
ON COALESCE(t1.k, '@$!(') = COALESCE(t2.k, '@$!(')
```
doesn't do a CrossJoin, despite being semantically near-identical.
I would greatly prefer to use the form `IS NOT DISTINCT FROM`, because then I don't have to come up with safe "junk values" that are known not to appear in the data, like '@$!(' here. 
I really don't see why Presto converts it to a CrossJoin. Hive and Spark have no problem doing an equi-join using the <=> operator, which has the exact same semantics as IS NOT DISTINCT FROM in Presto.

**Comment by findepi** [comment]
The Join operator uses equality comparison (because this is how 99% of joins are expressed) and, as you observed, it's "near identical" to `IS NOT DISTINCT FROM`, so the two cannot be used interchangeably.

Implementation-wise I see two possible approaches here:
- (easier) rewrite `Join on x IS NOT DISTINCT FROM y`  to `Join on coalesce(x, dummy) = coalesce(y dummy) and x IS NOT DISTINCT FROM y`
- (better) introduce direct support for `IS NOT DISTINCT FROM` in Join operator, as an (extended) equi-condition

I marked the latter option as better, because it avoids double comparison and also comparing `dummy` values that were originally in the source data with `dummy` values from `coalesce`. (if `dummy` is eg `0`, this would be a problem, but it can be arbitrary). However, the second approach is definitely much more involved.

cc @mbasmanova @sopel39 @kokosing

**Comment by findepi** [comment]
@yonillasky sure, i see your point. 

Just note -- if you generate those queries, you can actually solve / work around this very easily. Just do `Join on coalesce(x, dummy) = coalesce(y dummy) and x IS NOT DISTINCT FROM y` where `dummy` is arbitrary value (_can be present_ in the table, just pick something that is not very common. E.g. for an integer, 413 rather than 0). This should cause Presto use efficient hash-based Join, while results will be the same as for your original query.

**Comment by yonillasky** [comment]
Yes, that's what I did, however the code to support dummy values for arbitrary types was not entirely trivial:
```
    def _dummy_value(self, t: ColumnType) -> str:
        if isinstance(t, ColumnTypeBoolean):
            return 'FALSE'
        elif isinstance(t, ColumnTypeTinyint):
            return '-111'
        elif isinstance(t, ColumnTypeSmallint):
            return '-11111'
        elif isinstance(t, ColumnTypeInt):
            return '-111111111'
        elif isinstance(t, ColumnTypeBigint):
            return '-111111111111111111'
        elif isinstance(t, ColumnTypeFloat):
            return '-1.111e+11'
        elif isinstance(t, ColumnTypeDouble):
            return '-1.111e+111'
        elif isinstance(t, ColumnTypeString):
            return r"'~!@\N'"
        elif isinstance(t, ColumnTypeArray):
            return 'ARRAY[]'
        elif isinstance(t, ColumnTypeMap):
            return 'MAP(ARRAY[], ARRAY[])'
        elif isinstance(t, ColumnTypeStruct):
            values = ', '.join([self._dummy_value(f.type) for f in t.fields])
            schema = ', '.join([
                f"{f.name} {str(f.type).replace('string', 'varchar')}"
                for f in t.fields
            ])
            return f"CAST(ROW({values}) AS ROW({schema}))"
        else:
            # NOTE: Should be unreachable, we covered all types here
            raise TypeError(f"Can't generate dummy value for column type {t}")
```
and what I ended up doing is to put coalesce(t1.x, dummy) = coalesce(t2.x, dummy) in the join, and also generate a WHERE clause and check that t1.x IS NOT DISTINCT FROM t2.x

It did work out in the end. With that said, it's still somewhat odd that Hive/Spark support this type of equi-join and Presto doesn't.

**Comment by findepi** [comment]
@sopel39 observed that `coalesce(t1.x, dummy) = coalesce(t2.x, dummy)`-based approach isn't correct for floating point NaN.

**Comment by haozhun** [comment]
This is a duplicate of #6461 (and transitively #4075).

**Comment by findepi** [comment]
The only workaround I found turned out not to be correct (because there of NaNs, i.e. a non-NULL value that isn't equal to self).

Of course, this is "easily" fixable when generating a query, one just needs to generate `.x` in `coalesce(t1.x, dummy) = coalesce(t2.x, dummy)` into `CASE` replacing `NaN` with dummy.

I think workarond space is about exhausted, so i will close this in favor of https://github.com/prestodb/presto/issues/4075. (If it's not exhausted -- feel free to add more ideas in comments.)

**Comment by haozhun** [comment]
There is another potential workaround. "FULL OUTER JOIN"s can be written as "GROUP BY"s in some cases. I've met people who favor using GROUP BY for join when implementing SQL frameworks (due to its guarantee of not exploding rows in the presence of duplicate join keys due to data quality issues). Presto has fixed GROUP BY (#3568, since 0.201) recently to give it NOT DISTINCT FROM semantics.

```
SELECT id, arbitrary(a) a, arbitrary(b) b, ...
FROM ( 
  SELECT id, a, b, c, NULL, NULL, NULL FROM t1
  UNION ALL
  SELECT id, NULL, NULL, NULL, d, e, NULL FROM t2
  UNION ALL
  SELECT id, NULL, NULL, NULL, NULL, NULL, f FROM t3
)
GROUP BY id
